### ISSUE TITLE
consensus: how to handle blocks with invalid txs

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#2175]
Currently, we will commit blocks that may have invalid txs in them. This is done on purpose, to avoid requiring the block to be executed before consensus can go to work. This is a performance optimization, but comes at some UX cost, and led at least one user to raise an issue (#2162). 

Txs return results that include a status code, which can be non-zero indicating an invalid tx. We don't currently store this information anywhere in the blockchain itself (just indexed on disk). So long as we allow this, we should probably include a bit-array of valid/invalid txs in the header.

Another alternative is to just run CheckTx on all txs when we validate the block before prevoting. This would eliminate invalid txs (assuming CheckTx is implemented correctly), but would add some performance overhead.

Finally, we might want to consider a mechanism for detecting bad proposals and punishing for them, though this is a harder and less-immediate problem.

**Comment by ValarDragon** [comment]
I think it is advantageous to be able to include invalid txs in a block. The reason is, if we don't this may lead to validators electing certain message types due to them having a greater computational cost. (i.e. they fear if its invalid, they have more to lose)

However we should have to p2p layer reject invalid txs. I also agree with the bitarray in the header, that seems like a good decision. (However we may need to add slashing conditions for marking a tx as invalid when it actually should have been valid)

**Comment by ebuchman** [comment]
> The reason is, if we don't this may lead to validators electing certain message types due to them having a greater computational cost. (i.e. they fear if its invalid, they have more to lose)

Validators should be running CheckTx in their mempool before proposing a block of txs so this shouldnt be an issue. It's only a problem if a validator is trying to sneak txs into the block before validating them ...

> However we should have the p2p layer reject invalid txs.

It does at the mempool level. But if an invalid tx gets into a block via a malicious proposer, then we won't know until after the block is committed and executed, unless we do checktx before prevoting.

Note currently the expected logic is that the only way an invalid tx gets into a block is via a malicious proposer. Otherwise, all txs in a block should have had CheckTx run on them in the correct sequence. If CheckTx is implemented incorrectly, that's a separate issue.

> However we may need to add slashing conditions for marking a tx as invalid when it actually should have been valid

It would go in the next block and would be part of the validity check before prevoting for block H+1 to ensure that it has a correct results bit array for the txs in block H

**Comment by ValarDragon** [comment]
> Validators should be running CheckTx in their mempool before proposing a block of txs so this shouldnt be an issue. It's only a problem if a validator is trying to sneak txs into the block before validating them ...

I'm not sure. Suppose I am a validator solely concerned with maximizing my short term profit. Then I would want to propose txs with the maximum `fee / byte` ratio. (Or `fee / gas` in a computationally bounded chain) If an invalid tx has an extremely high fee and extremely large assoc. computational cost it may be advantageous for the validator to include it even though it was invalid. (Unless they get no reward for including it, in which case I completely agree)

> It does at the mempool level. But if an invalid tx gets into a block via a malicious proposer, then we won't know until after the block is committed and executed, unless we do checktx before prevoting.

Oh cool! Do we also lower the peers score in the address book if it was bad? (This may not make sense to do, since each block may change tx validity)

> It would go in the next block and would be part of the validity check before prevoting for block H+1 to ensure that it has a correct results bit array for the txs in block H

Doesn't that mess with instant finality? We want a block to be finalized as soon as it is published with 2/3rds votes right? We don't want to wait until the subsequent block to validate the previous block

**Comment by ebuchman** [comment]
> f an invalid tx has an extremely high fee and extremely large assoc. computational cost it may be advantageous for the validator to include it even though it was invalid. 

Right, I was just saying the code as written will always CheckTx before proposal. OF course validators could choose to include whatever they want, but they have to write that code.

Also note, CheckTx is supposed to be cheap, and txs passing CheckTx are not supposed to fail deliver tx. So the example might be a bit contrived.

> Oh cool! Do we also lower the peers score in the address book if it was bad? (This may not make sense to do, since each block may change tx validity)

No. It might be bad because we already have it too (eg. nonce is wrong now). Unless we change a lot about the mempool design, we can't punish for bad txs. Though, potentially, we could use ABCI codes more intelligently to eg. punish for txs that have encoding errors or something.

> Doesn't that mess with instant finality? We want a block to be finalized as soon as it is published with 2/3rds votes right? We don't want to wait until the subsequent block to validate the previous block

the finality technically comes when the commit is made - once you've seen the commit, you know its final. This is still the case, but in Tendermint the commit doesn't actually get included in the blockchain until the next block.

**Comment by ValarDragon** [comment]
> Also note, CheckTx is supposed to be cheap, and txs passing CheckTx are not supposed to fail deliver tx. So the example might be a bit contrived.

You're right. The only avenue to make the checkTx take forever is to make a large multisig with one invalid signature, but we can't take fees from a tx with an invalid signature. We probably want to add some sort of consensus parameter to bound the number of keys in a multisig, and to evaluate multisig pubkey count recursively. (As we want to support a multisig of multisig pubkeys)

However if we go with the original suggestion:
> Another alternative is to just run CheckTx on all txs when we validate the block before prevoting. 

I think it will be inevitable that CheckTx can't be both _fast_ and have no discrepancies between its result and the success of deliver tx. With that in mind, I think we should still allow txs that fail deliver tx. (Though I do agree with making it passing check tx, as long as check tx remains in its current fast variant. This only puts multisig txs at risk of being censored, due to higher risk at the same `fee per byte` or `fee per gas` levels)

> Though, potentially, we could use ABCI codes more intelligently to eg. punish for txs that have encoding errors or something.

This sounds like an awesome thing for us to do #postlaunch. Perhaps we should create another issue for this?

> This is still the case, but in Tendermint the commit doesn't actually get included in the blockchain until the next block. hence we have a concept of canonical and non-canonical commits. 

Thank you! I didn't know that the commit is published in the following block. You're right then, slashing conditions seem unnecessary.

**Comment by melekes** [comment]
> but would add some performance overhead.

can we measure this? so we know what're talking about

> we should probably include a bit-array of valid/invalid txs in the header.

sounds OK

**Comment by srmo** [comment]
I disagree. 
“Another alternative is to just run CheckTx on all txs when we validate the block before prevoting.”
From my understanding, it is perfectly valid to have two mutually exclusive tx A and B. Now depending on which node sees which tx first, you might end up with two validators X and Y where X had accepted A in the mempool, and Y accepted B. 
Now depending on who becomes the validator, either of them would reject A or B on a mempool precheck as checkTx always works against an ephemeral state since last state (at least that’s what we do). In such a scenario, it would need to be a special checkTx call which doesn’t assume an ephemeral state but always works against the last commited state. 

Please correct me if I’m wrong or if I should elaborate more on the use case.

**Comment by milosevic** [comment]
+1 for bit-array of valid/invalid txs in the header.

CheckTx is an application specific function. Therefore, some applications will just check format and possibly signature, some will check validity with respect to current stable state, some will check against ephemeral state. We need to be more clear what is the semantic we enforce from the Tendermint point of view. 
As always there are tradeoffs. Having a bit relaxed semantic (for example, a correct proposer will not include tx in in a proposal block if it does not pass checkTx) let us for example execute checkTx in parallel (as semantic does not say anything about order), and also make possible some multi-proposer optimisations at the consensus level (in order to increase throughput). Note that in this example we don't require checkTx calls to be sequential and to respect order of tx receptions, which is the semantic we enforce currently if I am not wrong. We should at least clarify what is the current semantic we are enforcing and the reasoning behind, so we can then potentially discuss alternatives.

**Comment by srmo** [comment]
You are right, that's why I was thinking about some "special" checkTx call.
In our scenario, "regular" checkTx will always work on an ephemeral state since last state transition.
But we could run a "special" checkTx against the last, commited state.

**Comment by ebuchman** [comment]
> checkTx always works against an ephemeral state since last state (at least that’s what we do). In such a scenario, it would need to be a special checkTx call which doesn’t assume an ephemeral state but always works against the last commited state.

That's right, and that's at least how we use CheckTx in apps. If we were to introduce CheckTx to block validity, perhaps it would happen on a new ABCI connection, specific for checking validity based on the last committed block. 

In any case, I think we should probably not try to introduce this new CheckTx now, and just ensure we track tx validity in headers.

**Comment by srmo** [comment]
@ebuchman ok.
Side question: something I'm still not really understanding, after the proposal, are the transactions propagated for the block propagated before running ABCI on each node? That's the only explanation I have as on how the "different mempools have different txs" scenario is solved, right?

**Comment by ebuchman** [comment]
When a validator is proposing, they send a proposal message with a signed hash of the block, and then separately gossip the pieces of the block, which include all the txs. But no one actually executes those txs until the block is committed. And its possible some nodes have never seen those txs before, while others have executed CheckTx on them in the mempool.

Does that answer your question?

**Comment by srmo** [comment]
> and then separately gossip the pieces of the block, which include all the txs

That's the answer. The "no one actually executes those" until ABCI runs in commit phase is understood.
Thanks :)

**Comment by ttmc** [comment]
This issue is very troubling because it breaks an assumption we've been making in our code. We won't be able to repair our code until this issue is resolved, either with the bit-array in the header or the option to CheckTx all transactions before prevoting.

What is the to-do list to resolve this issue (and issue #2162, which led to it)?

1. "include a bit-array of valid/invalid txs in the header."
1. "run CheckTx on all txs when we validate the block before prevoting." --- maybe make this configurable, and off by default
1. "explicitly punish proposers who propose bad blocks."

More?

Maybe those could be made into new issues, or maybe they could be made into explicit checkbox tasks under this issue. Punishing proposers doesn't seem as high a priority, so might be postponed until after launch, so maybe it makes sense to create a new issue for each of the listed tasks. I don't feel like I have the authority to do that.

**Comment by ebuchman** [comment]
Closing this for #2383 and #2384. 

> "explicitly punish proposers who propose bad blocks."

This can be done at the app level - the app knows if its returning a non-zero code in a DeliverTx

**Comment by krabradosty** [comment]
I can't find the information, how did you solve the problem? I recently started looking into the Tendermint and very concerned about how Tendermint treats the invalid transactions in blocks. I couldn't find anything in the documentation. Actually I think, that implement the "bit-array" is the best idea.
